reps/2022-12-15-optimizer-data.md

```diff
@@ -0,0 +1,308 @@
+
+Ray Datasets support both `Dataset` and `DatasetPipeline`.
+* `Dataset` is using eager mode by default. Lazy mode can be enabled by calling `Dataset.lazy()`.
+* `DatasetPipeline` is using lazy mode by default.
+
+#### 1.2. Proposed Change
+
+Change execution mode to lazy execution mode by default.
+
+Eager mode has the benefit when users are doing ad-hoc experiments or debugging. Users can call each operation line by line, to experiment or debug. However, eager mode has no benefit over lazy mode, when coming to production workload (non ad-hoc, running whole Python program in real cluster), with followed reasons:
+
+* When running the whole Python program, users do not debug the program line by line interactively.
+* Under eager mode, each operation is forced to be executed immediately upon calling. Optimizations to the whole execution plan are often impossible. This often leads to suboptimal performance.
+
+Lazy mode has proved to be effective and is the de-facto industry standard in the big data processing world (Spark, Presto/Trino, Hive). The major motivation with lazy mode is to build a plan optimizer to generate the optimal execution plan. It opens the door for more advanced techniques such as adaptive query execution, cost-based optimizer, pluggable execution back-end, etc. Lazy mode is the first step of the execution optimizer. Execution optimizer is the route to get us to optimal performance.
+
+### 2. Introduce execution plan optimizer
+
+#### 2.1. How do things work today?
+
+* `Dataset` has `ExecutionPlan`. 
+* `ExecutionPlan` has `BlockList` (as input) and a list of `Stage`s (as execution). 
+* When executing `ExecutionPlan`, `ExecutionPlan._optimize()` is called to get an optimal list of `Stage`s. 
+* After `_optimize()` is done, each `Stage` is executed one by one sequentially. First `Stage` reads input `BlockList`. All other `Stage` reads the output `BlockList` from the previous `Stage`, which is materialized in the Ray Object Store. 
+* `ExecutionPlan` supports partial execution (so-called snapshot), i.e. only executing some of `Stage`s but not all.
+
+#### 2.2. Proposed Change
+
+The current design works for now, but has a lot of room for improvement. Without a proper plan/query optimizer, it’s hard to extend and engineeringly impossible to maintain cleanly.
+
+##### 2.2.1. Interfaces
+
+```python
+class Optimizer():
+  """Abstract class for optimizer.
+
+  An optimizer transforms the plan, with a list of optimization rules and input statistics.
+  """
+
+  rules: List[Rule]
+
+  def optimize(plan: Plan, statistics) -> Plan:
+    for rule in rules:
+      plan = rule.apply(plan, statistics)
+    return plan 
+
+class LogicalOptimizer(Optimizer):
+  """Optimizer for logical plan.
+
+  An optimizer transforms the logical plan, with a list of optimization Rules and input statistics.
+  """
+
+  rules: List[Rule]
+
+  def optimize(plan: LogicalPlan, statistics) -> LogicalPlan:
```

Comment:
How do we go from LogicalPlan to PhysicalPlan, and what is the difference?

Might be good to have a figure and examples for a logical vs physical rule.

Reply:
> How do we go from LogicalPlan to PhysicalPlan, and what is the difference?

There would be a class/utility to convert `LogicalPlan` into `PhysicalPlan`. Given the current logic would be fairly straightforward, so leave this as implementation detail for now.

Added explanation for `LogicalPlan` and `PhysicalPlan` in Step 3 below.

> Might be good to have a figure and examples for a logical vs physical rule.

Added explanation and examples for logical and physical rule in Step 1 below.